pallets/pools/src/solution.rs

```diff
@@ -0,0 +1,649 @@
+		self
+	}
+
+	pub fn add_unhealthy(&mut self, add: UnhealthyState) -> &mut Self {
+		match self {
+			PoolState::Healthy => {
+				let mut states = Vec::new();
+				states.push(add);
+				*self = PoolState::Unhealthy(states);
+				self
+			}
+			PoolState::Unhealthy(states) => {
+				if !states.contains(&add) {
+					states.push(add);
+				}
+				self
+			}
+		}
+	}
+
+	pub fn rm_unhealthy(&mut self, rm: UnhealthyState) -> &mut Self {
+		match self {
+			PoolState::Healthy => {
+				let mut states = Vec::new();
+				states.push(rm);
+				*self = PoolState::Unhealthy(states);
+				self
+			}
+			PoolState::Unhealthy(states) => {
+				states.retain(|val| val != &rm);
+
+				if states.len() == 0 {
+					*self = PoolState::Healthy;
+				}
+				self
+			}
+		}
+	}
+}
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub enum UnhealthyState {
+	MaxReserveViolated,
+	MinRiskBufferViolated,
+}
+
+/// The solutions struct for epoch solution
+#[derive(Encode, Decode, Clone, Eq, RuntimeDebug, TypeInfo)]
+pub enum EpochSolution<Balance> {
+	Healthy(HealthySolution<Balance>),
+	Unhealthy(UnhealthySolution<Balance>),
+}
+
+impl<Balance> EpochSolution<Balance>
+where
+	Balance: Copy,
+{
+	pub fn healthy(&self) -> bool {
+		match self {
+			EpochSolution::Healthy(_) => true,
+			EpochSolution::Unhealthy(_) => false,
+		}
+	}
+
+	pub fn solution(&self) -> &[TrancheSolution] {
+		match self {
+			EpochSolution::Healthy(solution) => solution.solution.as_slice(),
+			EpochSolution::Unhealthy(solution) => solution.solution.as_slice(),
+		}
+	}
+}
+
+impl<Balance> PartialEq for EpochSolution<Balance>
+where
+	Balance: PartialEq,
+{
+	fn eq(&self, other: &Self) -> bool {
+		match self {
+			EpochSolution::Healthy(s_1) => match other {
+				EpochSolution::Healthy(s_2) => s_1 == s_2,
+				EpochSolution::Unhealthy(_) => false,
+			},
+			EpochSolution::Unhealthy(s_1) => match other {
+				EpochSolution::Healthy(_) => false,
+				EpochSolution::Unhealthy(s_2) => s_1 == s_2,
+			},
+		}
+	}
+}
+
+impl<Balance> PartialOrd for EpochSolution<Balance>
+where
+	Balance: PartialOrd,
+{
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		match self {
+			EpochSolution::Healthy(s_1) => match other {
+				EpochSolution::Healthy(s_2) => s_1.partial_cmp(s_2),
+				EpochSolution::Unhealthy(_) => Some(Ordering::Greater),
+			},
+			EpochSolution::Unhealthy(s_1) => match other {
+				EpochSolution::Healthy(_) => Some(Ordering::Less),
+				EpochSolution::Unhealthy(s_2) => s_1.partial_cmp(s_2),
+			},
+		}
+	}
+}
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct HealthySolution<Balance> {
+	pub solution: Vec<TrancheSolution>,
+	pub score: Balance,
+}
+
+impl<Balance> PartialOrd for HealthySolution<Balance>
+where
+	Balance: PartialOrd,
+{
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		let score_self = &self.score;
+		let score_other = &other.score;
+
+		Some(if score_self > score_other {
```

Comment:
Is this just `self.score.partial_cmp(&other.score)`?

Reply:
Probably, will change and see if tests go through.